piksi_tools/fileio.py

```diff
@@ -10,209 +11,416 @@
+
+    def __hash__(self):
+        return hash((self._seconds, self._millis))
+
+    def __repr__(self):
+        return "Time(s=%r,ms=%r)" % (self._seconds, self._millis)
+
+    def __add__(a, b):
+        new_time = (1000 * a._seconds) + (1000 * b._seconds) + a._millis + b._millis
+        return Time(seconds=(new_time / 1000), millis=(new_time % 1000))
+
+    def __eq__(a, b):
+        return a._seconds == b._seconds and a._millis == b._millis
+
+    def __ge__(a, b):
+        if a == b:
+            return True
+        return a > b
+
+    def __gt__(a, b):
+        if a._seconds < b._seconds:
+            return False
+        if a._seconds == b._seconds:
+            return a._millis > b._millis
+        return True
+
+
 class SelectiveRepeater(object):
-    """Selective repeater for SBP file transfers"""
+    """
+    Selective repeater for SBP file I/O requests
+
+    Fields
+    ----------
+    _pending_map : list(PendingRequest)
+      List (used as a map) of PendingRequest objects, used to track
+      outstanding requests.
+    _request_pool : Queue
+      Queue of available requests (recorded by index number)
+    _seqmap : dict(int,int)
+      Dictionary mapping SBP request sequence IDs to their corresponding
+      request index.
+    _batch_msgs : list
+      Collector for a batch of messages to be sent in one
+      buffer via the link
+    _last_check_time : Time
+      The last time we checked if any packets had expired
+    _expire_map : dict(Time, dict(PendingRequest, PendingRequest))
+      Dictionary which records the future time at which a request
+      will expire.
+
+    _msg_type : int
+      The message type we're currently sending
+    _link : Handler
+      The link over which we're sending data
+
+    _callback_thread : int
+      ID of the thread that we expect callbacks from
+    _link_thread : int
+      ID of the thread that handles link writes
+    """
 
     def __init__(self, link, msg_type, cb=None):
-        self.write_pool = [PendingWrite() for X in range(SBP_FILEIO_WINDOW_SIZE)]
-        self.pending = []
-        self.link = link
-        self.msg_type = msg_type
-        self.cb = cb
-        self.cb_thread = None
-        self.link_thread = None
+        """
+        Args
+        ---
+        link : Handler
+          Link over which messages will be sent.
+        msg_type:
+          The type of message being sent
+        """
+
+        self._link = link
+        self._msg_type = msg_type
+        self._callback = cb
+
+        self._seqmap = {}
+        self._batch_msgs = []
+        self._last_check_time = Time.now()
+        self._expire_map = defaultdict(dict)
+
+        self._init_fileio_config(SBP_FILEIO_WINDOW_SIZE, SBP_FILEIO_BATCH_SIZE)
+
+        self._callback_thread = None
+        self._link_thread = None
+
+        self._total_sends = 1.0
+        self._total_retries = 0
+
+        self._config_send_time = Time.now()
+        self._config_retry_time = Time.now()
```

Comment:
Should these be initialized to same instance/moment of time?

Reply:
Pretty sure this is OK just because of the order things are done in `self._config_receive()` but it probably makes more sense to setup the next expire time here and then also in the assignment of the value here https://github.com/swift-nav/piksi_tools/pull/968/files#diff-619782351bcba823c8db86598d3ca1b6R355.

Reply:
Fixed